velox/expression/ComplexProxyTypes.h

```diff
@@ -0,0 +1,218 @@
+
+ public:
+  ArrayProxy<V>(const ArrayProxy<V>&) = delete;
+
+  static bool constexpr provide_std_interface = CppToType<V>::isPrimitiveType &&
+      !std::is_same<Varchar, V>::value && !std::is_same<bool, V>::value;
+
+  // Add a new not null item to the array, increasing its size by 1.
+  FOLLY_ALWAYS_INLINE element_t& add_item() {
+    commitMostRecentChildItem();
+    auto index = valuesOffset_ + length_;
+    reserve(index + 1);
+    length_++;
+
+    if constexpr (!provide_std_interface) {
+      childWriter_->setOffset(index);
+      needCommit_ = true;
+      return childWriter_->current();
+    } else {
+      childWriter_->vector().setNull(index, false);
+      return childWriter_->data_[index];
+    }
+  }
+
+  // Add a new null item to the array.
+  FOLLY_ALWAYS_INLINE void add_null() {
+    commitMostRecentChildItem();
+    reserve(valuesOffset_ + length_ + 1);
+    childWriter_->vector().setNull(valuesOffset_ + length_, true);
+    length_++;
+    // Note: no need to commit the null item.
+  }
+
+  // Should be called by the user (VectorWriter) when writing is done to commit
+  // last item if needed.
+  void finalize() {
+    commitMostRecentChildItem();
+    // Downsize to the actual size used in the underlying vector.
+    // Some writer's logic depend on the previous size to append data.
+    childWriter_->vector().resize(valuesOffset_ + length_);
+  }
+
+  vector_size_t size() {
+    return length_;
+  }
+
+  // Functions below provides an std::like interface, and are enabled only when
+  // the array element is primitive that is not string or bool.
+
+  // Note: size is with respect to the current size of this array being written.
+  FOLLY_ALWAYS_INLINE typename std::enable_if<provide_std_interface>::type
+  resize(vector_size_t size) {
+    commitMostRecentChildItem();
+    reserve(size);
+    length_ = size;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(element_t value) {
+    auto& item = add_item();
+    item = value;
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(std::nullopt_t) {
+    add_null();
+  }
+
+  typename std::enable_if<provide_std_interface>::type FOLLY_ALWAYS_INLINE
+  push_back(const std::optional<element_t>& value) {
+    if (value) {
+      push_back(*value);
+    } else {
+      add_null();
+    }
+  }
+
+  typename std::
+      enable_if<provide_std_interface, PrimitiveNullableWriterProxy<V>>::type
+      operator[](vector_size_t index_) {
+    return PrimitiveNullableWriterProxy<V>{&childWriter_->vector(),
+                                           valuesOffset_ + index_};
+  }
+
+ private:
+  ArrayProxy<V>() {}
+
+  // Note: size is with respect to the current size of this array being written.
+  FOLLY_ALWAYS_INLINE void reserve(vector_size_t size) {
+    if (UNLIKELY(size > capacity_)) {
+      while (capacity_ < size) {
+        capacity_ = 2 * capacity_ + 1;
```

Comment:
Any thoughts on this strategy, obviously it will very quickly ramp up to give you adequate space, but I also feel it might be very wasteful. Some claim a more ideal growth rate is 1.5 : https://stackoverflow.com/questions/1100311/what-is-the-ideal-growth-rate-for-a-dynamically-allocated-array .

Reply:
I am not expert, i can use 1.5X as long as its growing expo.

Reply:
I too am not expert :), lot of research there claims 1.5 is good value but I will trust your research/judgement on this.